File: backend/app/schemas/catalog.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


class CatalogImportRow(BaseModel):
    title: str
    description: str | None = None
    brand: str | None = None
    color: str | None = None
    size: str | None = None
    material: str | None = None
    gender: str | None = None
    collection: str | None = None
    base_price: float | None = None
    currency: str | None = None
    category: str | None = None
    aliases: list[str] = Field(default_factory=list)
    variants: list[dict] = Field(default_factory=list)


class CatalogImportRequest(BaseModel):
    shop_id: UUID
    rows: list[CatalogImportRow] = Field(default_factory=list)
    source_format: str = "json"
    resume_job_id: UUID | None = None


class CatalogImportJobRead(BaseModel):
    id: UUID
    shop_id: UUID
    status: str
    source_format: str
    total_rows: int
    processed_rows: int
    failed_rows: int
    checkpoint: dict = Field(default_factory=dict)
    error_message: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    model_config = {"from_attributes": True}


class CatalogReindexRequest(BaseModel):
    shop_id: UUID
    product_ids: list[UUID] | None = None
    resume_job_id: UUID | None = None
    batch_size: int | None = Field(default=None, ge=1, le=500)


class CatalogReindexJobRead(BaseModel):
    job_id: UUID
    shop_id: UUID
    status: str
    total_products: int
    indexed_products: int
    checkpoint: dict = Field(default_factory=dict)


class ProductAliasRead(BaseModel):
    id: UUID
    alias_text: str
    language: str
    source: str
    confidence: float
    is_active: bool
    model_config = {"from_attributes": True}

    @field_validator("source", mode="before")
    @classmethod
    def coerce_source(cls, value: object) -> str:
        if hasattr(value, "value"):
            return str(value.value)
        return str(value)


class ProductNormalizedRead(BaseModel):
    id: UUID
    shop_id: UUID
    product_id: UUID
    normalized_title: str
    brand: str | None = None
    color: str | None = None
    size: str | None = None
    material: str | None = None
    gender: str | None = None
    collection: str | None = None
    synonym_candidates: list[str] = Field(default_factory=list)
    qdrant_point_id: str | None = None
    embedding_model: str | None = None
    last_normalized_at: datetime | None = None
    last_indexed_at: datetime | None = None
    aliases: list[ProductAliasRead] = Field(default_factory=list)
    model_config = {"from_attributes": True}


class CatalogProductListResponse(BaseModel):
    items: list[ProductNormalizedRead]
    total: int
    page: int
    page_size: int


class ProductAliasesPatchRequest(BaseModel):
    add: list[str] = Field(default_factory=list)
    remove: list[str] = Field(default_factory=list)
    language: str = "und"


class ProductAliasesPatchResponse(BaseModel):
    product_id: UUID
    aliases: list[ProductAliasRead]

File: frontend/src/types/catalog.ts
export interface ProductAlias {
  id: string;
  alias_text: string;
  language: string;
  source: string;
  confidence: number;
  is_active: boolean;
}

export interface ProductNormalized {
  id: string;
  shop_id: string;
  product_id: string;
  normalized_title: string;
  brand?: string | null;
  color?: string | null;
  size?: string | null;
  material?: string | null;
  gender?: string | null;
  collection?: string | null;
  synonym_candidates: string[];
  qdrant_point_id?: string | null;
  embedding_model?: string | null;
  last_normalized_at?: string | null;
  last_indexed_at?: string | null;
  aliases: ProductAlias[];
}

export interface CatalogProductListResponse {
  items: ProductNormalized[];
  total: number;
  page: number;
  page_size: number;
}

export interface CatalogImportJob {
  id: string;
  shop_id: string;
  status: string;
  source_format: string;
  total_rows: number;
  processed_rows: number;
  failed_rows: number;
  checkpoint: Record<string, unknown>;
  error_message?: string | null;
}

export interface CatalogImportRequest {
  shop_id: string;
  rows: Array<{
    title: string;
    description?: string;
    brand?: string;
    aliases?: string[];
    variants?: Array<{ color?: string; size?: string; sku?: string; stock_quantity?: number }>;
  }>;
}

export interface CatalogReindexRequest {
  shop_id: string;
  product_ids?: string[];
  batch_size?: number;
}

export interface ProductAliasesPatchRequest {
  add?: string[];
  remove?: string[];
  language?: string;
}
